libmscore/cmd.cpp

```diff
@@ -2848,6 +2848,124 @@ void Score::cmdAddGrace(NoteType graceType, int duration)
+///   clear measures, apply noBreak, set measureRepeatCount
+///   returns false if these measures won't work or user aborted
+//---------------------------------------------------------
+
+bool Score::makeMeasureRepeatGroup(Measure* firstMeasure, int numMeasures, int staffIdx)
+{
+    //
+    // check that sufficient measures exist, with equal durations
+    //
+    std::vector<Measure*> measures;
+    Measure* measure = firstMeasure;
+    for (int i = 1; i <= numMeasures; ++i) {
+        if (!measure || measure->ticks() != firstMeasure->ticks()) {
+            MScore::setError(INSUFFICIENT_MEASURES);
+            return false;
+        }
+        measures.push_back(measure);
+        measure = measure->nextMeasure();
+    }
+
+    //
+    // warn user if things will have to be deleted to make room for measure repeat
+    //
+    bool empty = true;
+    for (auto m : measures) {
+        if (m != measures.back()) {
+            if (m->endBarLineType() != BarLineType::NORMAL) {
+                empty = false;
+                break;
+            }
+        }
+        for (auto seg = m->first(); seg && empty; seg = seg->next()) {
+            if (seg->segmentType() & SegmentType::ChordRest) {
+                int strack = staffIdx * VOICES;
+                int etrack = strack + VOICES;
+                for (int track = strack; track < etrack; ++track) {
+                    Element* e = seg->element(track);
+                    if (e && !e->generated() && !e->isRest()) {
+                        empty = false;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    if (!empty) {
+        auto b = QMessageBox::warning(0, QObject::tr("Current contents of measures will be replaced"),
+                                      // QMessageBox titles aren't being shown, so include in message
+                                      QObject::tr("Current contents of measures will be replaced.")
+                                      + QObject::tr("\nContinue with inserting measure repeat?"),
+                                      QMessageBox::Cancel | QMessageBox::Ok,
+                                      QMessageBox::Ok);
+        if (b == QMessageBox::Cancel) {
+            return false;
+        }
+    }
+
+    //
+    // group measures and clear current contents
+    //
+
+    deselectAll();
+    int i = 1;
+    for (auto m : measures) {
+        select(m, SelectType::RANGE, staffIdx);
+        if (m->isMeasureRepeatGroup(staffIdx)) {
+            deleteItem(m->measureRepeatElement(staffIdx)); // reset measures related to an earlier MeasureRepeat
+        }
+        undoChangeMeasureRepeatCount(m, i++, staffIdx);
+        if (m != measures.front()) {
+            m->undoChangeProperty(Pid::REPEAT_START, false);
+        }
+        if (m != measures.back()) {
+            m->undoSetNoBreak(true);
+            deleteItem(const_cast<BarLine*>(m->endBarLine())); // also takes care of Pid::REPEAT_END
```

Comment:
Ive been guilty of using const_cast to get around things too, so I'm not necessarily one to talk, but - of all the things that maybe you shouldn't do with a const object, deleting it seems like it would be up there.  Does it perhaps make sense to reconsider the endBarLine method?

Reply:
Fair point, will do.

Reply:
Done